Refactor Inducing Point Selection to Use Allocator Classes for Enhanced Flexibility (#377)

yalsaffar opened the pull request:
# Description
This PR proposes a solution to issue #378 . Although the implementation is fully functional and tested, it is presented as an overall proposal, open for discussion and potential refinement.

### Key Changes
- **Refactor of `select_inducing_points` Function**:
  - **Previous Implementation**: Accepted `method` as a string (`"pivoted_chol"`, `"kmeans++"`, `"auto"`, `"sobol"`) and used conditional logic to select inducing points.
  - **New Implementation**: Now accepts an `InducingPointAllocator` instance, with `AutoAllocator` as the default if `allocator` is `None`. This approach allows users to directly pass allocator instances, aligning with the issue’s goal to enable flexible use of custom allocators like `GreedyImprovementReduction`.

- **New `inducing_point_allocators.py` File**:
  - Introduces classes `SobolAllocator`, `KMeansAllocator`, and `AutoAllocator`, all implementing the `InducingPointAllocator` interface from Botorch. This modularizes allocator logic, moving it out of `select_inducing_points` while following the established base class structure.

- **Modifications to Models and Example Files**:
  - Updated `gp_classification.py`, `monotonic_projection_gp.py`, `monotonic_rejection_gp.py`, `semi_p.py`, and `example_problems.py` to handle allocator class instances rather than string-based methods, improving overall consistency.
  - Added imports for the new allocator classes in `__init__.py` for cross-codebase accessibility.

- **Updated Tests**:
  - Adjusted tests in `test_semi_p.py`, `test_utils.py`, and `test_config.py` to work with allocator classes instead of the previous string-based structure.

### Additional Notes
This PR preserves most of the existing logic in `select_inducing_points` to keep changes minimal. I know further work is needed to confirm compatibility with additional Botorch allocators and to support advanced configurations using `from_config` for custom allocator setups. I’d love to hear your feedback on the overall approach before moving forward with these additional refinements.

crasanders reviewed (changes requested):
Looks good other than a few nits. Our internal tests indicated merge conflicts, so you probably need to rebase.

yalsaffar commented on aepsych/models/utils.py:
```diff
@@ -54,53 +55,55 @@ def compute_p_quantile(
+    # Ensure required inputs are provided for certain allocators
+    if isinstance(allocator, SobolAllocator):
+        assert bounds is not None, "Bounds must be provided for Sobol allocator!"
+    elif isinstance(allocator, (GreedyVarianceReduction, AutoAllocator)):
```
True, will fix that.

crasanders commented on tests/test_config.py:
```diff
@@ -1017,7 +1018,7 @@ def test_semip_config(self):
         self.assertTrue(model.inducing_size == 10)
         self.assertTrue(model.stim_dim == 1)
-        self.assertTrue(model.inducing_point_method == "sobol")
+        self.assertTrue(model.inducing_point_method == SobolAllocator)
```
Let's make a new file for InducingPointAllocator tests and check that each class can be correctly read from a config with its arguments (if any).

crasanders commented on aepsych/models/inducing_point_allocators.py:
```diff
@@ -0,0 +1,191 @@
+        return inducing_points   
+    
+    @classmethod
+    def from_config(cls, config:Config) -> 'SobolAllocator':
```
I thought about this more, and I think we should have the InducingPointAllocators read from configs themselves rather than receive inputs from the model. I think practically what this will look like is we need to register all of our InducingPointAllocators to work with configs, as well as the ones from botorch (see how we do this for botorch's acquisition functions). In the model's `from_config`, we'll get the InducingPointAllocator class and check if it has its own `from_config` method. If it does, we call it, if it doesn't, we can still try to instantiate it without any arguments. We could have some logic to try to infer what arguments should go into the constructor like we do for acquisition functions, but that code is pretty messy, so I don't know if that's a good idea. 

To maintain backwards compatibility, we can check if `inducing_point_method` is a string, and if it is, pass that string to `select_inducing_points`, but add a deprecation warning to that function that advises that you use the InducingPointAllocator classes directly.

yalsaffar commented:
### Changes

1. **Config-Driven Initialization for Allocators**:
   - Updated `inducing_point_allocators` to read parameters directly from the configuration file. Custom allocators (`SobolAllocator`, `KMeansAllocator`, `AutoAllocator`) now have a `from_config` method to allow instantiation with config-defined parameters.
   - For BoTorch allocators without a `from_config` method, the system attempts instantiation without additional parameters(similar to acquisitions).

2. **Enhanced Model `from_config` Handling**:
   - Models now check if an allocator class has a `from_config` method and use it if available, streamlining configuration-based initialization. 
   - Maintains backward compatibility: if `inducing_point_method` is provided as a string, it falls back to using the legacy approach in the `select_inducing_points` method.

3. **Updated `select_inducing_points` Function**:
   - `select_inducing_points` now supports both config-based allocator classes and string-based legacy methods (`"sobol"`, `"auto"`, etc.). When a legacy method is used, a `DeprecationWarning` is raised, guiding users to switch to the allocator classes.

4. **Testing Updates**:
   - Updated tests in `test_semi_p_config` within `test_config` to work with the new allocator structure.
   - Added `test_points_allocators.py` to validate:
     - Config-driven instantiation of allocators.
     - Bounds handling.
     - Legacy method support with deprecation warnings.
     - Compatibility with BoTorch allocators (`GreedyVarianceReduction`, `GreedyImprovementReduction`).

JasonKChow commented:
A new style for defining initializations from config just went in. Moving forward, we want to ensure all objects initialized to use the new ConfigurableMixin. Can you update to match the new spec?

ConfigurableMixin got updated here:
https://github.com/facebookresearch/aepsych/blob/main/aepsych/config.py#L430

You can see how I implemented it in parameters here:
https://github.com/facebookresearch/aepsych/blob/main/aepsych/transforms/parameters.py#L98

JasonKChow commented on aepsych/models/inducing_point_allocators.py:
```diff
@@ -0,0 +1,303 @@
+
+        # Run k-means++ on the unique inputs to select inducing points
+        inducing_points = torch.tensor(
+            kmeans2(unique_inputs.numpy(), num_inducing, minit="++")[0],
```
`unique_inputs.numpy()` fails on GPU. Need to force it to cpu first.
```suggestion
            kmeans2(unique_inputs.cpu().numpy(), num_inducing, minit="++")[0],
```

crasanders commented on aepsych/models/inducing_point_allocators.py:
```diff
@@ -0,0 +1,303 @@
+        """
+        if name is None:
+            name = cls.__name__
+        bounds = config.gettensor(name, "bounds")
```
I think we usually don't have bounds explicitly in the config, so we should construct them from lb/ub.

JasonKChow replied:
I think a utility function went in recently that'll produce 2D bounds. It should be in the root utils.py

JasonKChow replied:
Still a problem. Find the get_bounds function in transform and use it here. Ensure a test covers any case where this would break if get_bounds moved.

crasanders commented on aepsych/models/inducing_point_allocators.py:
```diff
@@ -0,0 +1,304 @@
+        return inducing_points
+
+    @classmethod
+    def from_config(
```
I think we can remove this, and all the other `from_config` methods, because these classes can just inherit the base implementation from the mixin, no?

crasanders reviewed (changes requested):
See my comment about the `from_config` implementations. I think once that is resolved, we can land this.

JasonKChow commented:
@yalsaffar Need rebase.

JasonKChow commented on aepsych/models/monotonic_rejection_gp.py:
Weirdly, I think monotonic rejection GP cannot have their inducing point method set via config. Add it to support your new allocators.

JasonKChow commented on aepsych/models/semi_p.py:
```diff
@@ -259,7 +261,7 @@ def __init__(
         inducing_size: Optional[int] = None,
         max_fit_time: Optional[float] = None,
-        inducing_point_method: str = "auto",
+        inducing_point_method: Optional[InducingPointAllocator] = None,
```
Default change.

JasonKChow commented on aepsych/models/semi_p.py:
```diff
@@ -279,11 +281,7 @@ def __init__(
-                If "pivoted_chol", selects points based on the pivoted Cholesky heuristic.
-                If "kmeans++", selects points by performing kmeans++ clustering on the training data.
-                If "auto", tries to determine the best method automatically.
+            inducing_point_method (InducingPointAllocator, optional): The method to use to select the inducing points. Defaults to AutoAllocator.
```
Default change.

JasonKChow commented on aepsych/models/semi_p.py:
```diff
@@ -315,6 +313,8 @@ def __init__(
         assert isinstance(
             likelihood, LinearBernoulliLikelihood
         ), "SemiP model only supports linear Bernoulli likelihoods!"
+        if inducing_point_method is None:
```
Default change.

JasonKChow commented on aepsych/models/semi_p.py:
```diff
@@ -528,7 +532,7 @@ def __init__(
         inducing_size: Optional[int] = None,
         max_fit_time: Optional[float] = None,
-        inducing_point_method: str = "auto",
+        inducing_point_method: Optional[InducingPointAllocator] = None,
```
Default change.

JasonKChow commented on aepsych/models/semi_p.py:
```diff
@@ -556,6 +560,8 @@ def __init__(
             optimizer_options (Dict[str, Any], optional): Optimizer options to pass to the SciPy optimizer during
                 fitting. Assumes we are using L-BFGS-B.
         """
+        if inducing_point_method is None:
```
Default change.

JasonKChow commented on aepsych/models/utils.py:
```diff
@@ -61,72 +65,85 @@ def compute_p_quantile(
 
 def select_inducing_points(
```
Consider carefully if this whole utility function is needed.  A lot of the logic here should probably be placed in AutoAllocator instead which can create new allocators within itself to manage auto selection.

JasonKChow commented on tests/models/test_gp_classification.py:
```diff
@@ -798,10 +798,21 @@ def test_extra_ask_warns(self):
         # verify that creating the model `from_config` or with `__init__` has the same hyperparams
 
-        m1 = GPClassificationModel(lb=[1, 2], ub=[3, 4])
+        m1 = GPClassificationModel(lb=[1, 2], ub=[3, 4], inducing_size=2)
```
Why was this needed, does default the default not work? Defaults should be able to take over if the argument is not set.

JasonKChow commented on tests/models/test_utils.py:
```diff
@@ -6,11 +6,17 @@
 # LICENSE file in the root directory of this source tree.
 
 import unittest
+from io import IOBase
```
Why is this needed?

JasonKChow commented on tests/models/test_utils.py:
```diff
@@ -79,24 +85,28 @@ def test_select_inducing_points(self):
             20,
         )
 
-        with self.assertRaises(AssertionError):
```
Assess whether the select_inducing_points function should e used at all anymore and consider replacing this check with a commensurate check that looks for invalid inputs (possibly for AutoAllocators).

JasonKChow commented on tests/test_points_allocators.py:
```diff
@@ -0,0 +1,159 @@
+            [par1]
+            par_type = continuous
+            lower_bound = 0.0
+            upper_bound = 1.0
```
Check if bounds are correct if a transformation is applied (easiest to check is `log_scale = True`).

JasonKChow replied:
It must be tested here, transforms are almost always applied (the Normalize transform), the log_scale is simply the most obvious when something goes wrong.

JasonKChow commented on tests/test_points_allocators.py:
```diff
@@ -0,0 +1,159 @@
+from botorch.utils.sampling import draw_sobol_samples
+
+
+class TestInducingPointAllocators(unittest.TestCase):
```
Need to test instantiating a full model and checking if the inducing points are correct: initially dummy but with the correct inducing_point_method set. Then after fitting once, inducing points are replaced with real inducing points.

JasonKChow commented on tests/test_points_allocators.py:
```diff
@@ -0,0 +1,159 @@
+    def test_greedy_variance_reduction_allocate_inducing_points(self):
+        # Mock data for testing
+        inputs = torch.rand(100, 2)
+        model = SingleTaskGP(inputs, torch.sin(inputs.sum(dim=-1)).unsqueeze(-1))
```
Should test with our models for completeness. See above for testing inducing points in a model. This might be a good place to do that.

JasonKChow reviewed (commented):
Unresolved comments.

yalsaffar commented on aepsych/models/gp_classification.py:
```diff
@@ -122,7 +125,7 @@ def __init__(
         )
         super().__init__(variational_strategy)
-
+        self.variational_strategy = variational_strategy
```
The main reason for saving this separately was that it is being called in `test_points_allocators.py`, where `model.variational_strategy.inducing_points` is retrieved and used to check if they match.

JasonKChow replied:
The test shouldn't require changing the spec of the model. Change the test and not the model. You can probably also access the objects by looking closely at where the generated inducing points actually go in the modules.
